HoRNDIS.cpp

```diff
@@ -582,182 +417,297 @@ bool HoRNDIS::createNetworkInterface() {
+   resources when the interface is sitting disabled and unused. */
 IOReturn HoRNDIS::enable(IONetworkInterface *netif) {
-	IONetworkMedium	*medium;
 	IOReturn rtn = kIOReturnSuccess;
-	
-	LOG(V_DEBUG, "enable from tid %p", current_thread());
+
+	LOG(V_DEBUG, "begin for thread_id=%lld", thread_tid(current_thread()));
+	ReentryLocker locker(this, fEnableDisableInProgress);
+	if (locker.isInterrupted()) {
+		LOG(V_ERROR, "Waiting interrupted");
+		return locker.getResult();
+	}
 
 	if (fNetifEnabled) {
-		LOG(V_ERROR, "already enabled?");
+		LOG(V_DEBUG, "Repeated call (thread_id=%lld), returning success",
+			thread_tid(current_thread()));
 		return kIOReturnSuccess;
 	}
-	
-	if (!allocateResources())
+
+	if (fCallbackCount != 0) {
+		LOG(V_ERROR, "Invalid state: fCallbackCount(=%d) != 0", fCallbackCount);
+		return kIOReturnError;
+	}
+
+	if (!allocateResources()) {
 		return kIOReturnNoMemory;
+	}
+
+	// Tell the other end to start transmitting.
+	if (!rndisSetPacketFilter(RNDIS_DEFAULT_FILTER)) {
+		goto bailout;
+	}
+
+	// We can now perform reads and writes between Network stack and USB device:
+	fReadyToTransfer = true;
 	
-	if (!fMediumDict)
-		if (!createMediumTables()) {
-			rtn = kIOReturnNoMemory;
+	// Kick off the read requests:
+	for (int i = 0; i < N_IN_BUFS; i++) {
+		pipebuf_t &inbuf = inbufs[i];
+		inbuf.comp.owner = this;
+		inbuf.comp.action = dataReadComplete;
+		inbuf.comp.parameter = &inbuf;
+
+		rtn = robustIO(fInPipe, &inbuf, (uint32_t)inbuf.mdp->getLength());
+		if (rtn != kIOReturnSuccess) {
+			LOG(V_ERROR, "Failed to start the first read: %08x\n", rtn);
 			goto bailout;
 		}
-	setCurrentMedium(IONetworkMedium::medium(kIOMediumEthernetAuto, 480 * 1000000));
-	
-	/* Kick off the first read. */
-	inbuf.comp.target = this;
-	inbuf.comp.action = dataReadComplete;
-	inbuf.comp.parameter = NULL;
-	
-	rtn = fInPipe->Read(inbuf.mdp, &inbuf.comp, NULL);
-	if (rtn != kIOReturnSuccess)
+		fCallbackCount++;
+	}
+
+	// Tell the world that the link is up...
+	if (!setLinkStatus(kIONetworkLinkActive | kIONetworkLinkValid,
+			getCurrentMedium(), getCurrentMedium()->getSpeed())) {
+		LOG(V_ERROR, "Cannot set link status");
+		rtn = kIOReturnError;
 		goto bailout;
+	}
 
-	/* Tell the world that the link is up... */
-	medium = IONetworkMedium::getMediumWithType(fMediumDict, kIOMediumEthernetAuto);
-	setLinkStatus(kIONetworkLinkActive | kIONetworkLinkValid, medium, 480 * 1000000);
-	
-	/* ... and then listen for packets! */
+	// ... and then listen for packets!
 	getOutputQueue()->setCapacity(TRANSMIT_QUEUE_SIZE);
 	getOutputQueue()->start();
 	LOG(V_DEBUG, "txqueue started");
-	
-	/* Tell the other end to start transmitting. */
-	if (!rndisSetPacketFilter(RNDIS_DEFAULT_FILTER))
-		goto bailout;
-	
-	/* Now we can say we're alive. */
+
+	// Now we can say we're alive.
 	fNetifEnabled = true;
-	
-	LOG(V_DEBUG, "done from tid %p", current_thread());
+	LOG(V_NOTE, "completed (thread_id=%lld): tethering interface '%s' "
```

Comment:
"RNDIS network interface" (not all RNDIS devices are tethering)

Reply:
done